### PR TITLE
Changed query string to unordered map

### PR DESCRIPTION
**Pull request by eneskuluk**
Changes variable query_ from string to unordered_map which uses key as string and value as string.

### CODE REVIEW

File: test/src/misc/test_uri.cc [hunk 1]
```diff
@@ -1,35 +1,49 @@
...
-  test_uri("hdfs://", "hdfs", "", 0, "", "");
-  test_uri("hdfs:///", "hdfs", "", 0, "/", "");
-  test_uri("hdfs:///tmp", "hdfs", "", 0, "/tmp", "");
-
-  test_uri("s3://s3-bucket/tmp", "s3", "s3-bucket", 0, "/tmp", "");
-
-  test_uri("gs://gcs-bucket/tmp", "gs", "gcs-bucket", 0, "/tmp", "");
-
-  test_uri("fdfdfd://dfdfd/fdfdf?fdf", "fdfdfd", "dfdfd", 0, "/fdfdf", "fdf");
+  test_uri("hdfs://oda-master:9000/tmp", "hdfs", "oda-master", 9000, "/tmp", std::vector<std::pair<std::string, std::string>>());
```
Comment (eneskuluk) [inline test/src/misc/test_uri.cc:25]: should i change the empty vectors to a single variable that is reused in the empty queries? like `std::vector<std::pair<std::string, std::string>> emptyVector` and use this in empty queries instead of using seperate vectors each time?

File: core/src/misc/uri.cc [hunk 2]
```diff
@@ -121,13 +120,39 @@ void uri::parse(const std::string& uri_s)
...
-    ++query_iter;
-    query_.assign(query_iter, end_iter);
+std::size_t startQuery = uri_s.find('?') + 1;
+  while(startQuery != std::string::npos && startQuery <= uri_s.length() - 2){
+    std::size_t endQuery = uri_s.find('=', startQuery);
+    std::size_t nextQuery = uri_s.find('&', startQuery);
+    // todo: handle edge cases
+    if(endQuery == std::string::npos || endQuery > nextQuery){
+      if(nextQuery == std::string::npos){
+        //throw std::system_error(EINVAL, std::generic_category(), "Query doesn't have value or not in proper format");
```
Comment (eneskuluk) [inline core/src/misc/uri.cc:135]: Should we throw error and terminate if query includes errors? currently it continues to parse even if it includes errors, it just skips to next parameter. Like following query 
`?query&anotherquery=someval&otherquery&` will result in vector having only one variable, key as anotherquery, value as someval with no errors. since initial and last 2 query parameters are missing/incorrect.

File: core/src/misc/uri.cc [hunk 3]
```diff
@@ -121,13 +120,39 @@ void uri::parse(const std::string& uri_s)
...
+      continue;
+    }  
+
+    std::string queryParam = uri_s.substr(startQuery,endQuery - startQuery);
+    startQuery = endQuery;
+    endQuery = nextQuery;
+    if(endQuery == std::string::npos)
+      endQuery = uri_s.length();
+    /*
+    handle edge cases later
```
Comment (eneskuluk) [inline core/src/misc/uri.cc:148]: there could be edge cases i forgot to include, tried to include that I could think of at that moment, will try to update If i think/encounter new ones.

File: core/include/misc/uri.h [hunk 4]
```diff
@@ -55,7 +56,7 @@ struct uri {
   std::string port_;
   int16_t nport_ = 0;
   std::string path_;
-  std::string query_;
+  std::vector<std::pair<std::string, std::string>> query_;
```
Comment (eneskuluk) [inline core/include/misc/uri.h:59]: Initially thought about using unordered set since retrieval would be faster If we are going to search keys when processing query, but since the size of the vector going to be very small, I think it is negligible right?

Comment (nalinigans) [reply core/include/misc/uri.h:59]: Actually using a set is a good idea as the code will be a lot cleaner for searching by keys.

Comment (nalinigans) [reply core/include/misc/uri.h]: Does not matter, unordered set or map is great. Would it be possible to change to `unordered map`?

Comment (eneskuluk) [reply core/include/misc/uri.h]: Sure, I already changed it. It is currently in the checks

File: core/src/misc/uri.cc (hunk 2 above)
Comment (nalinigans) [reply core/src/misc/uri.cc:135]: Yes, we need to throw exceptions for errors and catch the exception to return an error status(usually an integer of 0) in the caller.

Comment (nalinigans) [inline core/src/misc/uri.cc:135]: Yes, we need to throw exceptions for errors. The exception should be handled by the caller.

File: core/src/misc/uri.cc (hunk 3 above)
Comment (nalinigans) [reply core/src/misc/uri.cc:148]: See if you can create unit tests that would be edge cases.

File: test/src/misc/test_uri.cc [hunk 5]
```diff
@@ -1,35 +1,47 @@
...
+void test_uri(const std::string path_uri_str, const std::string protocol,  const std::string host, const int16_t nport, const std::string path, const std::unordered_map<std::string, std::string> query) {
   uri path_uri(path_uri_str);
 
   INFO("URI=" << path_uri_str);
   CHECK_THAT(path_uri.protocol(), Equals(protocol));
   CHECK_THAT(path_uri.host(), Equals(host));
   CHECK(path_uri.nport() == nport);
   CHECK_THAT(path_uri.path(), Equals(path));
-  CHECK_THAT(path_uri.query(), Equals(query));
+  CHECK(path_uri.query().empty() == query.empty());
```
Comment (nalinigans) [inline test/src/misc/test_uri.cc:13]: Should'nt this line be `if query.empty() CHECK(path_uri.query().empty() == query.empty());`

Comment (nalinigans) [reply test/src/misc/test_uri.cc:13]: Also, you can add these lines 
```
if (!query.empty()) {
  CHECK(path_uri.query.length() == query.length);
  for(auto pair: query){
     CHECK(path_uri.query().find(pair.first) != path_uri.query().end());
     CHECK_THAT(pair.second, Equals(path_uri.query()[pair.first]));
   }
}
```

**Review by nalinigans — CHANGES_REQUESTED** [review]
Please do add tests that throw exceptions in the query processing too. Thanks.

File: core/src/misc/uri.cc (hunk 2 above)
Comment (eneskuluk) [reply core/src/misc/uri.cc:135]: Is following enough: ` throw std::system_error(EINVAL, std::generic_category(), "Query is in incorrect format");`
What do you mean by caller? 
On testing side I did `REQUIRE_THROWS(test_uri("fdfdfd://dfdfd/fdfdf?firstQ=firstval&secondQ", "fdfdfd", "dfdfd", 0, "/fdfdf",emptyMap));`
Is this the correct way to do it?

Comment (eneskuluk) [reply core/src/misc/uri.cc:135]: query above should fail in the code right? currently it is skipping improperly formatted part picking correctly formatted parts. if a part of the query fails, should we end it completely?

Comment (nalinigans) [reply core/src/misc/uri.cc:135]: Caller in this context is the method calling/using the URI functionality. But, yes, this change with the testing side is enough.

File: test/src/misc/test_uri.cc (hunk 5 above)
Comment (eneskuluk) [reply test/src/misc/test_uri.cc:13]: I think it is better to check even if query is empty since path_uri's map could be not empty. So checking their lengths could be good to catch it. If query is empty, it won't go into the for loop anyway. Does the following looks good to you:
```
CHECK(path_uri.query.length() == query.length);
for(auto pair: query){
   CHECK(path_uri.query().find(pair.first) != path_uri.query().end());
   CHECK_THAT(pair.second, Equals(path_uri.query()[pair.first]));
 }
```

File: core/src/misc/uri.cc [hunk 6]
```diff
@@ -121,13 +120,33 @@ void uri::parse(const std::string& uri_s)
...
+  // No query available
+  if(uri_s.find('?') == std::string::npos || uri_s.find('=') == std::string::npos)
+    return;
 
-  std::string::const_iterator query_iter = find(path_iter, end_iter, '?');
-  path_.assign(path_iter, query_iter);
-
-  if (query_iter != end_iter) {
-    ++query_iter;
-    query_.assign(query_iter, end_iter);
```
Comment (nalinigans) [inline core/src/misc/uri.cc:130]: Just wondering if you could make this code a little bit concise using the existing code? Something on the lines of the following -
```
  if (query_iter != end_iter) {
    ++query_iter;
    query_.assign(query_iter, end_iter);
    char *save_ptr = query_.data();
    while (query_token_pair = strtok_r(query_.data(), "&", &save_ptr)) {
      key_length = query_token_pair.find("=");
      std::cout << "key=" <<  query_token_pair.substr(0, key_length-2) << " value=" << query_token_pair.substr(key_length, query_token_pair.length()-1) << std::endl;
    }
  }
```

File: core/src/misc/uri.cc [hunk 7]
```diff
@@ -121,13 +120,33 @@ void uri::parse(const std::string& uri_s)
...
+std::size_t startQuery = uri_s.find('?') + 1;
```
Comment (nalinigans) [inline core/src/misc/uri.cc:128]: You are starting to parse from the beginning that is not necessary. You should continue from `path_iter` in the original code. See the example code I posted with the original code.

**Review by nalinigans — CHANGES_REQUESTED** [review]
Just wondering if you could format the new code roughly with [Google Style](https://google.github.io/styleguide/cppguide.html)? @webberlh tells me VSCode has hooks for setting up this stye for formatting C/C++ source.

**Comment by webberlh** [comment]
>  @webberlh tells me VSCode has hooks for setting up this stye for formatting C/C++ source.
Specifically, I use the [clang-format](https://marketplace.visualstudio.com/items?itemName=xaver.clang-format) extension which can hook into your native `clang-format` binary for doing the formatting.

**Comment by eneskuluk** [comment]
> Just wondering if you could format the new code roughly with [Google Style](https://google.github.io/styleguide/cppguide.html)? @webberlh tells me VSCode has hooks for setting up this stye for formatting C/C++ source.

I now chose Google style formatting in cpp formatting in vs code settings. I'll format modified/added parts before committing. That should be enough for formatting wise right?

File: core/src/misc/uri.cc (hunk 7 above)
Comment (eneskuluk) [reply core/src/misc/uri.cc:128]: I see that makes sense, but when I used path_iter, it was returning iterator but I planned my code to work on indexes so it wasn't usable in my case. It starts looking for `?` from beginning, didn't think that would be a performance concern since there wouldn't be much characters until question mark. I'll check the example code thanks, I'll try to fix it.

File: core/src/misc/uri.cc (hunk 6 above)
Comment (eneskuluk) [reply core/src/misc/uri.cc:130]: Thanks for providing some sample code. I'll check if I can do it both readable and concise after handling the edge cases with following template.

**Review by nalinigans — APPROVED** [review]
LGTM

File: core/src/misc/uri.cc [hunk 8]
```diff
@@ -121,13 +120,23 @@ void uri::parse(const std::string& uri_s)
...
+    for (char* token = strtok_r(queryTemp.data(), "&", &save_ptr);
+         token != NULL; token = strtok_r(NULL, "&", &save_ptr)) {
+      char* search = strchr(token, '=');
+      if (search == NULL || (std::size_t)(search - token) == 0) {
+        throw std::system_error(EINVAL, std::generic_category(),
+                                "Query is in incorrect format");
+      }
+      std::string key(
+          std::string(token).substr(0, (std::size_t)(search - token)));
+      query_[key] = std::string(++search);
```
Comment (mlathara) [inline core/src/misc/uri.cc:138]: Do we need to decode from [url-encoded format](https://en.wikipedia.org/wiki/Query_string#URL_encoding) here?

Comment (nalinigans) [reply core/src/misc/uri.cc:138]: Not sure what you want done here, C++ is pretty agnostic to encodings, probably have to write our own decoders here. Also see [std::codecvt](https://en.cppreference.com/w/cpp/locale/codecvt) for some not so great support. My suggestion is to get the upstream components to pass already decoded uri strings to this layer.

Comment (nalinigans) [reply core/src/misc/uri.cc:138]: arrow uses `https://github.com/uriparser/uriparser` for parsing URIs

**Review by mlathara — APPROVED** [review]
lgtm
